kimchi/cairo/src/runner/bytecode.rs

```diff
@@ -0,0 +1,106 @@
+// TODO(querolita):
+// - be able to index memory with any type (u256 in particular) not only usize
+
+use crate::runner::word::CairoWord;
+
+/// This data structure stores the memory of the program
+pub struct CairoBytecode {
+    /// length of the public memory
+    publen: u64,
+    /// full memory vector, None if non initialized
+    pub vect: Vec<Option<CairoWord>>,
+}
+
+impl CairoBytecode {
+    /// Create a new memory structure from a vector of u64
+    pub fn new(input: Vec<i128>) -> CairoBytecode {
+        // Initialized with the public memory (compiled instructions only)
+        // starts intentionally with a zero word for ease of testing
+        let mut aux = vec![0];
+        aux.extend(input);
+        CairoBytecode {
+            publen: (aux.len() - 1) as u64,
+            vect: aux.into_iter().map(|i| Some(CairoWord::new(i))).collect(),
+        }
+    }
+
+    /// Get size of the public memory
+    pub fn public(&self) -> u64 {
+        self.publen
+    }
+
+    /// Get size of the full memory including dummy 0th entry
+    pub fn size(&self) -> u64 {
+        (self.vect.len()) as u64
+    }
+
+    /// Enlarges memory with enough additional None slots if necessary before writing or reading
+    fn enlarge(&mut self, index: u64) {
+        // if you want to access an index of the memory but its size is less or equal than this
+        if self.size() <= index {
+            // you will need to extend the vector with enough spaces (taking into account that
+            // vectors start by index 0 and size starts in 1)
+            let additional = index - self.size() + 1;
```

Comment:
For this, I think it makes sense to stick to the second option you're suggesting. Otherwise I would be moving the task of checking space to the `read` and `write` methods instead, and I prefer to have it in the same `resize` function. However, it may make sense to call it `check_or_resize` or something like that...